tesseract/tesseract_scene_graph/include/tesseract_scene_graph/resource_locator.h

```diff
@@ -0,0 +1,65 @@
+#ifndef TESSERACT_SCENE_GRAPH_RESOURCE_LOCATOR_H
+#define TESSERACT_SCENE_GRAPH_RESOURCE_LOCATOR_H
+
+#include <tesseract_common/resource.h>
+#include <tesseract_scene_graph/allowed_collision_matrix.h>
+#include <tesseract_scene_graph/graph.h>
+#include <tesseract_scene_graph/parser/srdf_parser.h>
+#include <functional>
+
+namespace tesseract_scene_graph
+{
+// ResourceLocator abstract class
+class ResourceLocator
+{
+public:
+  using Ptr = std::shared_ptr<ResourceLocator>;
+  using ConstPtr = std::shared_ptr<const ResourceLocator>;
+
+  virtual tesseract_common::Resource::Ptr locateResource(const std::string& url) = 0;
+};
+
+// SimpleResourceLocator using a function to resolve filenames
+class SimpleResourceLocator : public ResourceLocator
+{
+public:
+  using Ptr = std::shared_ptr<SimpleResourceLocator>;
+  using ConstPtr = std::shared_ptr<const SimpleResourceLocator>;
+
+  using ResourceLocatorFn = std::function<std::string(const std::string&)>;
+
+  SimpleResourceLocator(ResourceLocatorFn locator_function);
+
+  virtual tesseract_common::Resource::Ptr locateResource(const std::string& url) override;
+
+protected:
+  ResourceLocatorFn locator_function_;
+};
+
+// SimpleLocatedResource for file
+class SimpleLocatedResource : public tesseract_common::Resource
+{
+public:
+  using Ptr = std::shared_ptr<SimpleLocatedResource>;
+  using ConstPtr = std::shared_ptr<const SimpleLocatedResource>;
+
+  SimpleLocatedResource(const std::string& url, const std::string& filename);
+
+  virtual bool isFile() override;
+
+  virtual std::string getUrl() override;
+
+  virtual std::string getFilePath() override;
```

Comment:
I just have concern that this may need to be used within a const function which would not be allowed since these are not const. 

I would expect that the three function isFile(), getUrl() and getFilePath() should be able to be const?

I can see where getResourceContents and getResourceContentStream but I believe in the case of caching data it is permitted to make these function const and use mutable on the cached object along with mutex. What are your thoughts?

Reply:
Resolving the resource may also result in cached value like the real filename or real url. This would be stored in the `ResourceLocator`. The actual data would be cached in the `Resource`, and the `ResourceLocator` may cache the retrieved data in case it is requested again.

Reply:
@gavanderhoorn and @johnwason Do either you see an issue with making these const and leverage mutable when caching?